{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO, AlignIO\n",
    "import os, itertools, subprocess\n",
    "import pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CLD_CONTINUED',\n",
       " 'CLD_DUMPED',\n",
       " 'CLD_EXITED',\n",
       " 'CLD_TRAPPED',\n",
       " 'DirEntry',\n",
       " 'EX_CANTCREAT',\n",
       " 'EX_CONFIG',\n",
       " 'EX_DATAERR',\n",
       " 'EX_IOERR',\n",
       " 'EX_NOHOST',\n",
       " 'EX_NOINPUT',\n",
       " 'EX_NOPERM',\n",
       " 'EX_NOUSER',\n",
       " 'EX_OK',\n",
       " 'EX_OSERR',\n",
       " 'EX_OSFILE',\n",
       " 'EX_PROTOCOL',\n",
       " 'EX_SOFTWARE',\n",
       " 'EX_TEMPFAIL',\n",
       " 'EX_UNAVAILABLE',\n",
       " 'EX_USAGE',\n",
       " 'F_LOCK',\n",
       " 'F_OK',\n",
       " 'F_TEST',\n",
       " 'F_TLOCK',\n",
       " 'F_ULOCK',\n",
       " 'MutableMapping',\n",
       " 'NGROUPS_MAX',\n",
       " 'O_ACCMODE',\n",
       " 'O_APPEND',\n",
       " 'O_ASYNC',\n",
       " 'O_CLOEXEC',\n",
       " 'O_CREAT',\n",
       " 'O_DIRECT',\n",
       " 'O_DIRECTORY',\n",
       " 'O_DSYNC',\n",
       " 'O_EXCL',\n",
       " 'O_LARGEFILE',\n",
       " 'O_NDELAY',\n",
       " 'O_NOATIME',\n",
       " 'O_NOCTTY',\n",
       " 'O_NOFOLLOW',\n",
       " 'O_NONBLOCK',\n",
       " 'O_RDONLY',\n",
       " 'O_RDWR',\n",
       " 'O_RSYNC',\n",
       " 'O_SYNC',\n",
       " 'O_TRUNC',\n",
       " 'O_WRONLY',\n",
       " 'POSIX_FADV_DONTNEED',\n",
       " 'POSIX_FADV_NOREUSE',\n",
       " 'POSIX_FADV_NORMAL',\n",
       " 'POSIX_FADV_RANDOM',\n",
       " 'POSIX_FADV_SEQUENTIAL',\n",
       " 'POSIX_FADV_WILLNEED',\n",
       " 'PRIO_PGRP',\n",
       " 'PRIO_PROCESS',\n",
       " 'PRIO_USER',\n",
       " 'P_ALL',\n",
       " 'P_NOWAIT',\n",
       " 'P_NOWAITO',\n",
       " 'P_PGID',\n",
       " 'P_PID',\n",
       " 'P_WAIT',\n",
       " 'PathLike',\n",
       " 'RTLD_DEEPBIND',\n",
       " 'RTLD_GLOBAL',\n",
       " 'RTLD_LAZY',\n",
       " 'RTLD_LOCAL',\n",
       " 'RTLD_NODELETE',\n",
       " 'RTLD_NOLOAD',\n",
       " 'RTLD_NOW',\n",
       " 'R_OK',\n",
       " 'SCHED_BATCH',\n",
       " 'SCHED_FIFO',\n",
       " 'SCHED_IDLE',\n",
       " 'SCHED_OTHER',\n",
       " 'SCHED_RESET_ON_FORK',\n",
       " 'SCHED_RR',\n",
       " 'SEEK_CUR',\n",
       " 'SEEK_END',\n",
       " 'SEEK_SET',\n",
       " 'ST_APPEND',\n",
       " 'ST_MANDLOCK',\n",
       " 'ST_NOATIME',\n",
       " 'ST_NODEV',\n",
       " 'ST_NODIRATIME',\n",
       " 'ST_NOEXEC',\n",
       " 'ST_NOSUID',\n",
       " 'ST_RDONLY',\n",
       " 'ST_RELATIME',\n",
       " 'ST_SYNCHRONOUS',\n",
       " 'ST_WRITE',\n",
       " 'TMP_MAX',\n",
       " 'WCONTINUED',\n",
       " 'WCOREDUMP',\n",
       " 'WEXITED',\n",
       " 'WEXITSTATUS',\n",
       " 'WIFCONTINUED',\n",
       " 'WIFEXITED',\n",
       " 'WIFSIGNALED',\n",
       " 'WIFSTOPPED',\n",
       " 'WNOHANG',\n",
       " 'WNOWAIT',\n",
       " 'WSTOPPED',\n",
       " 'WSTOPSIG',\n",
       " 'WTERMSIG',\n",
       " 'WUNTRACED',\n",
       " 'W_OK',\n",
       " 'XATTR_CREATE',\n",
       " 'XATTR_REPLACE',\n",
       " 'XATTR_SIZE_MAX',\n",
       " 'X_OK',\n",
       " '_Environ',\n",
       " '__all__',\n",
       " '__builtins__',\n",
       " '__cached__',\n",
       " '__doc__',\n",
       " '__file__',\n",
       " '__loader__',\n",
       " '__name__',\n",
       " '__package__',\n",
       " '__spec__',\n",
       " '_execvpe',\n",
       " '_exists',\n",
       " '_exit',\n",
       " '_fspath',\n",
       " '_fwalk',\n",
       " '_get_exports_list',\n",
       " '_putenv',\n",
       " '_spawnvef',\n",
       " '_unsetenv',\n",
       " '_wrap_close',\n",
       " 'abc',\n",
       " 'abort',\n",
       " 'access',\n",
       " 'altsep',\n",
       " 'chdir',\n",
       " 'chmod',\n",
       " 'chown',\n",
       " 'chroot',\n",
       " 'close',\n",
       " 'closerange',\n",
       " 'confstr',\n",
       " 'confstr_names',\n",
       " 'cpu_count',\n",
       " 'ctermid',\n",
       " 'curdir',\n",
       " 'defpath',\n",
       " 'device_encoding',\n",
       " 'devnull',\n",
       " 'dup',\n",
       " 'dup2',\n",
       " 'environ',\n",
       " 'environb',\n",
       " 'error',\n",
       " 'execl',\n",
       " 'execle',\n",
       " 'execlp',\n",
       " 'execlpe',\n",
       " 'execv',\n",
       " 'execve',\n",
       " 'execvp',\n",
       " 'execvpe',\n",
       " 'extsep',\n",
       " 'fchdir',\n",
       " 'fchmod',\n",
       " 'fchown',\n",
       " 'fdatasync',\n",
       " 'fdopen',\n",
       " 'fork',\n",
       " 'forkpty',\n",
       " 'fpathconf',\n",
       " 'fsdecode',\n",
       " 'fsencode',\n",
       " 'fspath',\n",
       " 'fstat',\n",
       " 'fstatvfs',\n",
       " 'fsync',\n",
       " 'ftruncate',\n",
       " 'fwalk',\n",
       " 'get_blocking',\n",
       " 'get_exec_path',\n",
       " 'get_inheritable',\n",
       " 'get_terminal_size',\n",
       " 'getcwd',\n",
       " 'getcwdb',\n",
       " 'getegid',\n",
       " 'getenv',\n",
       " 'getenvb',\n",
       " 'geteuid',\n",
       " 'getgid',\n",
       " 'getgrouplist',\n",
       " 'getgroups',\n",
       " 'getloadavg',\n",
       " 'getlogin',\n",
       " 'getpgid',\n",
       " 'getpgrp',\n",
       " 'getpid',\n",
       " 'getppid',\n",
       " 'getpriority',\n",
       " 'getresgid',\n",
       " 'getresuid',\n",
       " 'getsid',\n",
       " 'getuid',\n",
       " 'getxattr',\n",
       " 'initgroups',\n",
       " 'isatty',\n",
       " 'kill',\n",
       " 'killpg',\n",
       " 'lchown',\n",
       " 'linesep',\n",
       " 'link',\n",
       " 'listdir',\n",
       " 'listxattr',\n",
       " 'lockf',\n",
       " 'lseek',\n",
       " 'lstat',\n",
       " 'major',\n",
       " 'makedev',\n",
       " 'makedirs',\n",
       " 'minor',\n",
       " 'mkdir',\n",
       " 'mkfifo',\n",
       " 'mknod',\n",
       " 'name',\n",
       " 'nice',\n",
       " 'open',\n",
       " 'openpty',\n",
       " 'pardir',\n",
       " 'path',\n",
       " 'pathconf',\n",
       " 'pathconf_names',\n",
       " 'pathsep',\n",
       " 'pipe',\n",
       " 'pipe2',\n",
       " 'popen',\n",
       " 'posix_fadvise',\n",
       " 'posix_fallocate',\n",
       " 'pread',\n",
       " 'preadv',\n",
       " 'putenv',\n",
       " 'pwrite',\n",
       " 'pwritev',\n",
       " 'read',\n",
       " 'readlink',\n",
       " 'readv',\n",
       " 'register_at_fork',\n",
       " 'remove',\n",
       " 'removedirs',\n",
       " 'removexattr',\n",
       " 'rename',\n",
       " 'renames',\n",
       " 'replace',\n",
       " 'rmdir',\n",
       " 'scandir',\n",
       " 'sched_get_priority_max',\n",
       " 'sched_get_priority_min',\n",
       " 'sched_getaffinity',\n",
       " 'sched_getparam',\n",
       " 'sched_getscheduler',\n",
       " 'sched_param',\n",
       " 'sched_rr_get_interval',\n",
       " 'sched_setaffinity',\n",
       " 'sched_setparam',\n",
       " 'sched_setscheduler',\n",
       " 'sched_yield',\n",
       " 'sendfile',\n",
       " 'sep',\n",
       " 'set_blocking',\n",
       " 'set_inheritable',\n",
       " 'setegid',\n",
       " 'seteuid',\n",
       " 'setgid',\n",
       " 'setgroups',\n",
       " 'setpgid',\n",
       " 'setpgrp',\n",
       " 'setpriority',\n",
       " 'setregid',\n",
       " 'setresgid',\n",
       " 'setresuid',\n",
       " 'setreuid',\n",
       " 'setsid',\n",
       " 'setuid',\n",
       " 'setxattr',\n",
       " 'spawnl',\n",
       " 'spawnle',\n",
       " 'spawnlp',\n",
       " 'spawnlpe',\n",
       " 'spawnv',\n",
       " 'spawnve',\n",
       " 'spawnvp',\n",
       " 'spawnvpe',\n",
       " 'st',\n",
       " 'stat',\n",
       " 'stat_result',\n",
       " 'statvfs',\n",
       " 'statvfs_result',\n",
       " 'strerror',\n",
       " 'supports_bytes_environ',\n",
       " 'supports_dir_fd',\n",
       " 'supports_effective_ids',\n",
       " 'supports_fd',\n",
       " 'supports_follow_symlinks',\n",
       " 'symlink',\n",
       " 'sync',\n",
       " 'sys',\n",
       " 'sysconf',\n",
       " 'sysconf_names',\n",
       " 'system',\n",
       " 'tcgetpgrp',\n",
       " 'tcsetpgrp',\n",
       " 'terminal_size',\n",
       " 'times',\n",
       " 'times_result',\n",
       " 'truncate',\n",
       " 'ttyname',\n",
       " 'umask',\n",
       " 'uname',\n",
       " 'uname_result',\n",
       " 'unlink',\n",
       " 'unsetenv',\n",
       " 'urandom',\n",
       " 'utime',\n",
       " 'wait',\n",
       " 'wait3',\n",
       " 'wait4',\n",
       " 'waitid',\n",
       " 'waitid_result',\n",
       " 'waitpid',\n",
       " 'walk',\n",
       " 'write',\n",
       " 'writev']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(os)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n"
     ]
    }
   ],
   "source": [
    "genbank = [\"BDIP_ARG.gb\", \"BPUB_RS.gb\", \"Bothrops_jararaca-NC_030760.1.gb\", \"Bothrops_pubescens-NC_039648.1.gb\", \"Bothrops_diporus-NC_039649.1.gb\"]\n",
    "print(len(genbank))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CDS\n",
    "For CDS's, the \"gene\" field provides an standardized gene code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDIP_ARG.gb \n",
      "\n",
      "ND1\n",
      "ND2\n",
      "COX1\n",
      "COX2\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "ND3\n",
      "ND4L\n",
      "ND4\n",
      "ND5\n",
      "ND6\n",
      "CYTB\n",
      "----------------------------\n",
      "\n",
      "BPUB_RS.gb \n",
      "\n",
      "ND1\n",
      "ND2\n",
      "COX1\n",
      "COX2\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "ND3\n",
      "ND4L\n",
      "ND4\n",
      "ND5\n",
      "ND6\n",
      "CYTB\n",
      "----------------------------\n",
      "\n",
      "Bothrops_jararaca-NC_030760.1.gb \n",
      "\n",
      "ND1\n",
      "ND2\n",
      "COX1\n",
      "COX2\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "ND3\n",
      "ND4L\n",
      "ND4\n",
      "ND5\n",
      "ND6\n",
      "CYTB\n",
      "----------------------------\n",
      "\n",
      "Bothrops_pubescens-NC_039648.1.gb \n",
      "\n",
      "ND1\n",
      "ND2\n",
      "COX1\n",
      "COX2\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "ND3\n",
      "ND4L\n",
      "ND4\n",
      "ND5\n",
      "ND6\n",
      "CYTB\n",
      "----------------------------\n",
      "\n",
      "Bothrops_diporus-NC_039649.1.gb \n",
      "\n",
      "ND1\n",
      "ND2\n",
      "COX1\n",
      "COX2\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "ND3\n",
      "ND4L\n",
      "ND4\n",
      "ND5\n",
      "ND6\n",
      "CYTB\n",
      "----------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "wanted_features = [\"CDS\", \"tRNA\", \"rRNA\"]\n",
    "recovered_features = list()\n",
    "for gb in genbank:\n",
    "    print(gb, \"\\n\")\n",
    "    for i in SeqIO.parse(gb, \"genbank\"):\n",
    "        organism = i.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "        for feature in i.features:\n",
    "            if feature.type == \"CDS\":\n",
    "            #if feature.qualifiers.get('gene') == [\"COX1\"]:\n",
    "                #print(gb)\n",
    "                print(feature.qualifiers.get('gene')[0])\n",
    "    print(\"----------------------------\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# tRNAS\n",
    "For the tRNAs, the field \"product\" should be used:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDIP_ARG.gb \n",
      "\n",
      "tRNA-Phe\n",
      "tRNA-Val\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "tRNA-Lys\n",
      "tRNA-Gly\n",
      "tRNA-Arg\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "tRNA-Glu\n",
      "tRNA-Thr\n",
      "----------------------------\n",
      "\n",
      "BPUB_RS.gb \n",
      "\n",
      "tRNA-Phe\n",
      "tRNA-Val\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "tRNA-Lys\n",
      "tRNA-Gly\n",
      "tRNA-Arg\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "tRNA-Glu\n",
      "tRNA-Thr\n",
      "tRNA-Pro\n",
      "----------------------------\n",
      "\n",
      "Bothrops_jararaca-NC_030760.1.gb \n",
      "\n",
      "tRNA-Phe\n",
      "tRNA-Val\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "tRNA-Lys\n",
      "tRNA-Gly\n",
      "tRNA-Arg\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "tRNA-Glu\n",
      "tRNA-Thr\n",
      "----------------------------\n",
      "\n",
      "Bothrops_pubescens-NC_039648.1.gb \n",
      "\n",
      "tRNA-Phe\n",
      "tRNA-Val\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "tRNA-Lys\n",
      "tRNA-Gly\n",
      "tRNA-Arg\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "tRNA-Glu\n",
      "tRNA-Thr\n",
      "tRNA-Pro\n",
      "----------------------------\n",
      "\n",
      "Bothrops_diporus-NC_039649.1.gb \n",
      "\n",
      "tRNA-Phe\n",
      "tRNA-Val\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "tRNA-Lys\n",
      "tRNA-Gly\n",
      "tRNA-Arg\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "tRNA-Glu\n",
      "tRNA-Thr\n",
      "----------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "wanted_features = [\"CDS\", \"tRNA\", \"rRNA\"]\n",
    "recovered_features = list()\n",
    "for gb in genbank:\n",
    "    print(gb, \"\\n\")\n",
    "    for i in SeqIO.parse(gb, \"genbank\"):\n",
    "        organism = i.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "        for feature in i.features:\n",
    "            if feature.type == \"tRNA\":\n",
    "            #if feature.qualifiers.get('product') == [\"tRNA-Ser\"]:\n",
    "                print(feature.qualifiers.get('product')[0])\n",
    "    print(\"----------------------------\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rRNAs\n",
    "For the rRNAs, there is variation among gene names (BDIP_ARG and BPUB_RS). However, the \"product\" field could be used to provide an valid id for these features:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDIP_ARG.gb \n",
      "\n",
      "['small subunit ribosomal RNA']\n",
      "['large subunit ribosomal RNA']\n",
      "----------------------------\n",
      "\n",
      "BPUB_RS.gb \n",
      "\n",
      "['small subunit ribosomal RNA']\n",
      "['large subunit ribosomal RNA']\n",
      "----------------------------\n",
      "\n",
      "Bothrops_jararaca-NC_030760.1.gb \n",
      "\n",
      "['12S ribosomal RNA']\n",
      "['16S ribosomal RNA']\n",
      "----------------------------\n",
      "\n",
      "Bothrops_pubescens-NC_039648.1.gb \n",
      "\n",
      "['12S ribosomal RNA']\n",
      "['16S ribosomal RNA']\n",
      "----------------------------\n",
      "\n",
      "Bothrops_diporus-NC_039649.1.gb \n",
      "\n",
      "['12S ribosomal RNA']\n",
      "['16S ribosomal RNA']\n",
      "----------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "wanted_features = [\"CDS\", \"tRNA\", \"rRNA\"]\n",
    "recovered_features = list()\n",
    "for gb in genbank:\n",
    "    print(gb, \"\\n\")\n",
    "    for i in SeqIO.parse(gb, \"genbank\"):\n",
    "        organism = i.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "        for feature in i.features:\n",
    "            if feature.type == \"rRNA\":\n",
    "            #if feature.qualifiers.get('product') == [\"tRNA-Ser\"]:\n",
    "                print(feature.qualifiers.get('product'))\n",
    "                #print(feature.qualifiers.get('product')[0])\n",
    "    print(\"----------------------------\\n\")\n",
    "                      \n",
    "    ''' if feature.type in wanted_features:\n",
    "                seq = feature.location.extract(i.seq)\n",
    "                if feature.type in recovered_features:\n",
    "                    feature.type = \"{}2\".format(feature.type)\n",
    "                print(feature.qualifiers.get('product'))\n",
    "                recovered_features.append(feature.qualifiers.get('product')[0])\n",
    "                feature_type = feature.qualifiers.get('product')[0]\n",
    "                feature = \">{}-{}\\n{}\\n\".format(feature_type, organism, seq)\n",
    "                #print(feature)'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene_order\n",
    "This piece of code could be useful for generating synteny figures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDIP_ARG.gb \n",
      "\n",
      "tRNA-Phe\n",
      "small subunit ribosomal RNA\n",
      "tRNA-Val\n",
      "large subunit ribosomal RNA\n",
      "ND1\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "ND2\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "COX1\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "COX2\n",
      "tRNA-Lys\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "tRNA-Gly\n",
      "ND3\n",
      "tRNA-Arg\n",
      "ND4L\n",
      "ND4\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "ND5\n",
      "ND6\n",
      "tRNA-Glu\n",
      "CYTB\n",
      "tRNA-Thr\n",
      "----------------------------\n",
      "\n",
      "BPUB_RS.gb \n",
      "\n",
      "tRNA-Phe\n",
      "small subunit ribosomal RNA\n",
      "tRNA-Val\n",
      "large subunit ribosomal RNA\n",
      "ND1\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "ND2\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "COX1\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "COX2\n",
      "tRNA-Lys\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "tRNA-Gly\n",
      "ND3\n",
      "tRNA-Arg\n",
      "ND4L\n",
      "ND4\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "ND5\n",
      "ND6\n",
      "tRNA-Glu\n",
      "CYTB\n",
      "tRNA-Thr\n",
      "tRNA-Pro\n",
      "----------------------------\n",
      "\n",
      "Bothrops_jararaca-NC_030760.1.gb \n",
      "\n",
      "tRNA-Phe\n",
      "12S ribosomal RNA\n",
      "tRNA-Val\n",
      "16S ribosomal RNA\n",
      "ND1\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "ND2\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "COX1\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "COX2\n",
      "tRNA-Lys\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "tRNA-Gly\n",
      "ND3\n",
      "tRNA-Arg\n",
      "ND4L\n",
      "ND4\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "ND5\n",
      "ND6\n",
      "tRNA-Glu\n",
      "CYTB\n",
      "tRNA-Thr\n",
      "----------------------------\n",
      "\n",
      "Bothrops_pubescens-NC_039648.1.gb \n",
      "\n",
      "tRNA-Phe\n",
      "12S ribosomal RNA\n",
      "tRNA-Val\n",
      "16S ribosomal RNA\n",
      "ND1\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "ND2\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "COX1\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "COX2\n",
      "tRNA-Lys\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "tRNA-Gly\n",
      "ND3\n",
      "tRNA-Arg\n",
      "ND4L\n",
      "ND4\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "ND5\n",
      "ND6\n",
      "tRNA-Glu\n",
      "CYTB\n",
      "tRNA-Thr\n",
      "tRNA-Pro\n",
      "----------------------------\n",
      "\n",
      "Bothrops_diporus-NC_039649.1.gb \n",
      "\n",
      "tRNA-Phe\n",
      "12S ribosomal RNA\n",
      "tRNA-Val\n",
      "16S ribosomal RNA\n",
      "ND1\n",
      "tRNA-Ile\n",
      "tRNA-Pro\n",
      "tRNA-Leu\n",
      "tRNA-Gln\n",
      "tRNA-Met\n",
      "ND2\n",
      "tRNA-Trp\n",
      "tRNA-Ala\n",
      "tRNA-Asn\n",
      "tRNA-Cys\n",
      "tRNA-Tyr\n",
      "COX1\n",
      "tRNA-Ser\n",
      "tRNA-Asp\n",
      "COX2\n",
      "tRNA-Lys\n",
      "ATP8\n",
      "ATP6\n",
      "COX3\n",
      "tRNA-Gly\n",
      "ND3\n",
      "tRNA-Arg\n",
      "ND4L\n",
      "ND4\n",
      "tRNA-His\n",
      "tRNA-Ser\n",
      "tRNA-Leu\n",
      "ND5\n",
      "ND6\n",
      "tRNA-Glu\n",
      "CYTB\n",
      "tRNA-Thr\n",
      "----------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "wanted_features = [\"CDS\", \"tRNA\", \"rRNA\"]\n",
    "recovered_features = list()\n",
    "for gb in genbank:\n",
    "    print(gb, \"\\n\")\n",
    "    for i in SeqIO.parse(gb, \"genbank\"):\n",
    "        organism = i.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "        for feature in i.features:\n",
    "            if feature.type == \"CDS\":\n",
    "            #if feature.qualifiers.get('product') == [\"tRNA-Ser\"]:\n",
    "                print(feature.qualifiers.get('gene')[0])\n",
    "            elif feature.type == \"tRNA\":\n",
    "                print(feature.qualifiers.get('product')[0])\n",
    "            elif feature.type == \"rRNA\":\n",
    "                print(feature.qualifiers.get('product')[0])\n",
    "    print(\"----------------------------\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Now, for the real thing:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDIP_ARG.gb \n",
      "\n",
      "BPUB_RS.gb \n",
      "\n",
      "Bothrops_jararaca-NC_030760.1.gb \n",
      "\n",
      "Bothrops_pubescens-NC_039648.1.gb \n",
      "\n",
      "Bothrops_diporus-NC_039649.1.gb \n",
      "\n"
     ]
    }
   ],
   "source": [
    "fasta_set = set()\n",
    "\n",
    "def feature_parser(genbank):\n",
    "    all_features = dict()\n",
    "    for gb in genbank:\n",
    "        recovered_features = list()\n",
    "        print(gb, \"\\n\")\n",
    "        gb_handle = SeqIO.read(gb, \"genbank\")\n",
    "        organism = gb_handle.annotations.get(\"organism\").replace(\" \", \"_\")\n",
    "        all_features[organism] = {}\n",
    "        for feature in gb_handle.features:\n",
    "            gene_name = ''\n",
    "            seq = feature.location.extract(gb_handle.seq)\n",
    "            if feature.type == \"CDS\":\n",
    "                #continue\n",
    "                #print(feature.qualifiers.get('gene')[0])\n",
    "                gene_name = feature.qualifiers.get('gene')[0]\n",
    "            #elif feature.type == \"tRNA\":\n",
    "                #continue\n",
    "                #print(feature.qualifiers.get('product')[0])\n",
    "             #   gene_name = feature.qualifiers.get('product')[0]\n",
    "            #elif feature.type == \"rRNA\":\n",
    "            #    if feature.qualifiers.get('product')[0].startswith('small'):\n",
    "            #        feature.qualifiers['product'] = ['12S ribosomal RNA']\n",
    "            #    if feature.qualifiers.get('product')[0].startswith('large'):\n",
    "            #        feature.qualifiers['product'] = ['16S ribosomal RNA']\n",
    "                #print(feature.qualifiers.get('product')[0].split(\" \")[0],'\\n',seq)\n",
    "                #gene_name = feature.qualifiers.get('product')[0].split(\" \")[0]\n",
    "            if gene_name != '':\n",
    "                if gene_name in recovered_features:\n",
    "                    gene_name = \"{}2\".format(gene_name)\n",
    "                recovered_features.append(gene_name)\n",
    "                feature_dict = {gene_name : seq}\n",
    "                all_features[organism].update(feature_dict)\n",
    "                \n",
    "                #with open(fasta_name, \"a+\") as fasta:\n",
    "                    #fasta_set.add(fasta_name)\n",
    "                    #fasta.write(\">{}_{}\\n{}\\n\".format(organism,gene_name,seq))\n",
    "                #print(header)\n",
    "                #print(recovered_features)\n",
    "                #print(len(fasta_set))\n",
    "                #print(fasta_set)\n",
    "                #print('>{} {}'.format(header, organism))\n",
    "        #print(\"----------------------------\\n\")\n",
    "    return(all_features)\n",
    "\n",
    "feature_dict = feature_parser(genbank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDIP_ARG\n",
      "1 ND1\n",
      "2 ND2\n",
      "3 COX1\n",
      "4 COX2\n",
      "5 ATP8\n",
      "6 ATP6\n",
      "7 COX3\n",
      "8 ND3\n",
      "9 ND4L\n",
      "10 ND4\n",
      "11 ND5\n",
      "12 ND6\n",
      "13 CYTB\n",
      "BPUB_RS\n",
      "1 ND1\n",
      "2 ND2\n",
      "3 COX1\n",
      "4 COX2\n",
      "5 ATP8\n",
      "6 ATP6\n",
      "7 COX3\n",
      "8 ND3\n",
      "9 ND4L\n",
      "10 ND4\n",
      "11 ND5\n",
      "12 ND6\n",
      "13 CYTB\n",
      "Bothrops_jararaca\n",
      "1 ND1\n",
      "2 ND2\n",
      "3 COX1\n",
      "4 COX2\n",
      "5 ATP8\n",
      "6 ATP6\n",
      "7 COX3\n",
      "8 ND3\n",
      "9 ND4L\n",
      "10 ND4\n",
      "11 ND5\n",
      "12 ND6\n",
      "13 CYTB\n",
      "Bothrops_pubescens\n",
      "1 ND1\n",
      "2 ND2\n",
      "3 COX1\n",
      "4 COX2\n",
      "5 ATP8\n",
      "6 ATP6\n",
      "7 COX3\n",
      "8 ND3\n",
      "9 ND4L\n",
      "10 ND4\n",
      "11 ND5\n",
      "12 ND6\n",
      "13 CYTB\n",
      "Bothrops_diporus\n",
      "1 ND1\n",
      "2 ND2\n",
      "3 COX1\n",
      "4 COX2\n",
      "5 ATP8\n",
      "6 ATP6\n",
      "7 COX3\n",
      "8 ND3\n",
      "9 ND4L\n",
      "10 ND4\n",
      "11 ND5\n",
      "12 ND6\n",
      "13 CYTB\n"
     ]
    }
   ],
   "source": [
    "for species in feature_dict.keys():\n",
    "    print(species)\n",
    "    for n, i in enumerate(feature_dict[species].keys(), 1):\n",
    "        print(n, i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BDIP_ARG': {'ND1': Seq('ATAATCTCAACCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()),\n",
       "  'ND2': Seq('ATTAACCCCACATCACTAATAACCATTACTACCAGCATCATCTTAAGTACCACA...TTT', IUPACAmbiguousDNA()),\n",
       "  'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()),\n",
       "  'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()),\n",
       "  'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()),\n",
       "  'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...ATA', IUPACAmbiguousDNA()),\n",
       "  'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()),\n",
       "  'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()),\n",
       "  'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...ATA', IUPACAmbiguousDNA()),\n",
       "  'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCCCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()),\n",
       "  'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTATCCTGGTGCTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()),\n",
       "  'CYTB': Seq('ATGCCCCACCAACATTTACTCGCCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())},\n",
       " 'BPUB_RS': {'ND1': Seq('ATAATCTCAGTCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()),\n",
       "  'ND2': Seq('ATTAACCCCACATCACTAATAACCATTACAACCAGCATCATCTTAAGTACCACA...TAT', IUPACAmbiguousDNA()),\n",
       "  'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()),\n",
       "  'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()),\n",
       "  'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()),\n",
       "  'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...ATA', IUPACAmbiguousDNA()),\n",
       "  'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()),\n",
       "  'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()),\n",
       "  'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...ATA', IUPACAmbiguousDNA()),\n",
       "  'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCTCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()),\n",
       "  'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTGTCTTGGTGTTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()),\n",
       "  'CYTB': Seq('ATGCCCCACCAACATTTACTCACCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())},\n",
       " 'Bothrops_jararaca': {'ND1': Seq('ATAATCCCAATCCTACTCAACATCATGAACCCCTTACTTTACATCCTACCCATC...AAT', IUPACAmbiguousDNA()),\n",
       "  'ND2': Seq('ATTAACCCCACATCACTAATAACCATCACAATTAGCATCATCTTAAGCACTACC...TAT', IUPACAmbiguousDNA()),\n",
       "  'COX1': Seq('GTGTACATTACCCGTTGACTGTTTTCAACCAACCACAAAGATATTGGAACCCTA...AGA', IUPACAmbiguousDNA()),\n",
       "  'COX2': Seq('ATGCCCTACGCCACCCAACTATCACTACAAGAAGCCCTAAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()),\n",
       "  'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTTCAGTGTACCTGTGAACCTGACTG...TAA', IUPACAmbiguousDNA()),\n",
       "  'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCAGAAATCATCCTCACCCCA...TAA', IUPACAmbiguousDNA()),\n",
       "  'COX3': Seq('ATGACCCACCAGCTACATCAATACCATATAGTCGATCCTAGCCCATGACCCCTG...CGT', IUPACAmbiguousDNA()),\n",
       "  'ND3': Seq('ATCAGCCTAATACTTCTCATCATCTTAGCCACAACAACAGCAACCTTCCTATAT...AGT', IUPACAmbiguousDNA()),\n",
       "  'ND4L': Seq('ATGGAACTAATCAAAACCATTCTAGCCATGACCTTCATAACCACCATTCTTAGC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND4': Seq('ATGCTCAAAGTCCTCATCATAACAGCAATACTAATCCCAACAATCTTAACTCTA...AGA', IUPACAmbiguousDNA()),\n",
       "  'ND5': Seq('ATGGACCTAATAACACCAACTATCACTCTAACAATCTTCCTCTCCCTTACCGCC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND6': Seq('ATGTTTTTAGTAGAGTACTTGCTTTATTTTATTTTAGTGTTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()),\n",
       "  'CYTB': Seq('ATGCCCCACCAACACTTGCTCACCCTATTCAACCTACTCCCCGTAGGGTCAAAT...AAT', IUPACAmbiguousDNA())},\n",
       " 'Bothrops_pubescens': {'ND1': Seq('ATAATCTCAGTCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()),\n",
       "  'ND2': Seq('ATTAACCCCACATCACTAATAACCATTACAACCAGCATCATCTTAAGTACCACA...TAT', IUPACAmbiguousDNA()),\n",
       "  'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()),\n",
       "  'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()),\n",
       "  'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()),\n",
       "  'ATP6': Seq('ATGCCCACACACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...TAA', IUPACAmbiguousDNA()),\n",
       "  'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()),\n",
       "  'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()),\n",
       "  'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCTCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()),\n",
       "  'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTATCTTGGCGCTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()),\n",
       "  'CYTB': Seq('ATGCCCCACCAACATTTACTCGCCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())},\n",
       " 'Bothrops_diporus': {'ND1': Seq('ATAATCTCAGTCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()),\n",
       "  'ND2': Seq('ATTAACCCCACATCACTAATAACCGTTACAACCAGCATCATCTTAAGTACCACA...TAT', IUPACAmbiguousDNA()),\n",
       "  'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()),\n",
       "  'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()),\n",
       "  'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()),\n",
       "  'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...TAA', IUPACAmbiguousDNA()),\n",
       "  'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()),\n",
       "  'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()),\n",
       "  'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCTCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()),\n",
       "  'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()),\n",
       "  'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTGTCTTGGTGTTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()),\n",
       "  'CYTB': Seq('ATGCCCCACCAACATTTACTCGCCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())}}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ND1': Seq('ATAATCTCAACCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()), 'ND2': Seq('ATTAACCCCACATCACTAATAACCATTACTACCAGCATCATCTTAAGTACCACA...TTT', IUPACAmbiguousDNA()), 'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()), 'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()), 'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()), 'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...ATA', IUPACAmbiguousDNA()), 'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()), 'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()), 'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...ATA', IUPACAmbiguousDNA()), 'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCCCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()), 'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()), 'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTATCCTGGTGCTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()), 'CYTB': Seq('ATGCCCCACCAACATTTACTCGCCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())}\n",
      "{'ND1': Seq('ATAATCTCAGTCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()), 'ND2': Seq('ATTAACCCCACATCACTAATAACCATTACAACCAGCATCATCTTAAGTACCACA...TAT', IUPACAmbiguousDNA()), 'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()), 'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()), 'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()), 'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...ATA', IUPACAmbiguousDNA()), 'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()), 'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()), 'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...ATA', IUPACAmbiguousDNA()), 'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCTCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()), 'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()), 'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTGTCTTGGTGTTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()), 'CYTB': Seq('ATGCCCCACCAACATTTACTCACCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())}\n",
      "{'ND1': Seq('ATAATCCCAATCCTACTCAACATCATGAACCCCTTACTTTACATCCTACCCATC...AAT', IUPACAmbiguousDNA()), 'ND2': Seq('ATTAACCCCACATCACTAATAACCATCACAATTAGCATCATCTTAAGCACTACC...TAT', IUPACAmbiguousDNA()), 'COX1': Seq('GTGTACATTACCCGTTGACTGTTTTCAACCAACCACAAAGATATTGGAACCCTA...AGA', IUPACAmbiguousDNA()), 'COX2': Seq('ATGCCCTACGCCACCCAACTATCACTACAAGAAGCCCTAAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()), 'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTTCAGTGTACCTGTGAACCTGACTG...TAA', IUPACAmbiguousDNA()), 'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCAGAAATCATCCTCACCCCA...TAA', IUPACAmbiguousDNA()), 'COX3': Seq('ATGACCCACCAGCTACATCAATACCATATAGTCGATCCTAGCCCATGACCCCTG...CGT', IUPACAmbiguousDNA()), 'ND3': Seq('ATCAGCCTAATACTTCTCATCATCTTAGCCACAACAACAGCAACCTTCCTATAT...AGT', IUPACAmbiguousDNA()), 'ND4L': Seq('ATGGAACTAATCAAAACCATTCTAGCCATGACCTTCATAACCACCATTCTTAGC...TAA', IUPACAmbiguousDNA()), 'ND4': Seq('ATGCTCAAAGTCCTCATCATAACAGCAATACTAATCCCAACAATCTTAACTCTA...AGA', IUPACAmbiguousDNA()), 'ND5': Seq('ATGGACCTAATAACACCAACTATCACTCTAACAATCTTCCTCTCCCTTACCGCC...TAA', IUPACAmbiguousDNA()), 'ND6': Seq('ATGTTTTTAGTAGAGTACTTGCTTTATTTTATTTTAGTGTTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()), 'CYTB': Seq('ATGCCCCACCAACACTTGCTCACCCTATTCAACCTACTCCCCGTAGGGTCAAAT...AAT', IUPACAmbiguousDNA())}\n",
      "{'ND1': Seq('ATAATCTCAGTCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()), 'ND2': Seq('ATTAACCCCACATCACTAATAACCATTACAACCAGCATCATCTTAAGTACCACA...TAT', IUPACAmbiguousDNA()), 'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()), 'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()), 'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()), 'ATP6': Seq('ATGCCCACACACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...TAA', IUPACAmbiguousDNA()), 'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()), 'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()), 'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...TAA', IUPACAmbiguousDNA()), 'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCTCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()), 'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()), 'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTATCTTGGCGCTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()), 'CYTB': Seq('ATGCCCCACCAACATTTACTCGCCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())}\n",
      "{'ND1': Seq('ATAATCTCAGTCCTACTCAACATCATAAACCCCTTACTTTTCATCCTGCCCATC...AAT', IUPACAmbiguousDNA()), 'ND2': Seq('ATTAACCCCACATCACTAATAACCGTTACAACCAGCATCATCTTAAGTACCACA...TAT', IUPACAmbiguousDNA()), 'COX1': Seq('GTGTACATCACCCGTTGACTATTTTCAACCAACCACAAAGACATTGGAACCCTA...AGA', IUPACAmbiguousDNA()), 'COX2': Seq('ATGCCCTACGCCACCCAGCTTTCACTACAAGAAGCCCTTAGCCCAACAATAGAA...AAT', IUPACAmbiguousDNA()), 'ATP8': Seq('ATGCCACAACTAGACATCGTACATATCCTTACAGTATACCTGTGAGCCTGATTA...TAA', IUPACAmbiguousDNA()), 'ATP6': Seq('ATGAACATAAACATATTTGAACAATTTATAAGCCCTGAAATCCTCCTCACCCCA...TAA', IUPACAmbiguousDNA()), 'COX3': Seq('ATGACCCACCAGCTACATCAATACCACATAGTCGACCCCAGCCCATGACCTCTG...CGT', IUPACAmbiguousDNA()), 'ND3': Seq('ATTAACCTGACACTTCTCATCATCATAGCCACAGCAACAGCAACCCTCCTATAC...AGT', IUPACAmbiguousDNA()), 'ND4L': Seq('ATGGAACTAATCAAAATCATTCTAACCATGGCCTTCATAACCACCATTCTTAGC...TAA', IUPACAmbiguousDNA()), 'ND4': Seq('ATGCTCAAAATCCTTATCATAACAGCAATACTAGCTCCAACAATCCTAACCCTG...AGA', IUPACAmbiguousDNA()), 'ND5': Seq('ATGGACCTAATAACACCCACTATTGTTCTAGCAATTTTCCTCTCCCTCACCCTC...TAA', IUPACAmbiguousDNA()), 'ND6': Seq('ATGTTTATAGTAGAGTACCTGCTTTATTTTGTCTTGGTGTTGGTGTTTTTTGGT...AGG', IUPACAmbiguousDNA()), 'CYTB': Seq('ATGCCCCACCAACATTTACTCGCCCTATTCAACCTACTTCCAGTAGGATCGAAC...ACT', IUPACAmbiguousDNA())}\n"
     ]
    }
   ],
   "source": [
    "for gene in feature_dict.values():\n",
    "        print(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ND1': 5, 'ND2': 5, 'COX1': 5, 'COX2': 5, 'ATP8': 5, 'ATP6': 5, 'COX3': 5, 'ND3': 5, 'ND4L': 5, 'ND4': 5, 'ND5': 5, 'ND6': 5, 'CYTB': 5}\n"
     ]
    }
   ],
   "source": [
    "def remove_feature_from_dict(feature):\n",
    "    for gene in feature_dict.values():\n",
    "        if feature in gene.keys():\n",
    "            del gene[feature]\n",
    "        \n",
    "def count_features(feature_dict):\n",
    "    counter = dict()\n",
    "    for gene in feature_dict.values():\n",
    "        for i in gene.keys():\n",
    "            if i not in counter:\n",
    "                counter[i] = 1\n",
    "            else:\n",
    "                counter[i] += 1\n",
    "    return(counter)\n",
    "\n",
    "def remove_missing_features(feature_dict):                \n",
    "    counter = count_features(feature_dict)\n",
    "    print(counter)\n",
    "    for key, value in counter.items():\n",
    "        if value < len(feature_dict):\n",
    "            print(key)\n",
    "            remove_feature_from_dict(key)\n",
    "    return(feature_dict)\n",
    "\n",
    "feature_dict = remove_missing_features(feature_dict)\n",
    "\n",
    "    \n",
    "#new_feature_dict = remove_missing_features(feature_dict)\n",
    "        \n",
    "#print(len(feature_dict))\n",
    "#print(count_features)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['ND1', 'ND2', 'COX1', 'COX2', 'ATP8', 'ATP6', 'COX3', 'ND3', 'ND4L', 'ND4', 'ND5', 'ND6', 'CYTB']) \n",
      "\n",
      "dict_keys(['ND1', 'ND2', 'COX1', 'COX2', 'ATP8', 'ATP6', 'COX3', 'ND3', 'ND4L', 'ND4', 'ND5', 'ND6', 'CYTB']) \n",
      "\n",
      "dict_keys(['ND1', 'ND2', 'COX1', 'COX2', 'ATP8', 'ATP6', 'COX3', 'ND3', 'ND4L', 'ND4', 'ND5', 'ND6', 'CYTB']) \n",
      "\n",
      "dict_keys(['ND1', 'ND2', 'COX1', 'COX2', 'ATP8', 'ATP6', 'COX3', 'ND3', 'ND4L', 'ND4', 'ND5', 'ND6', 'CYTB']) \n",
      "\n",
      "dict_keys(['ND1', 'ND2', 'COX1', 'COX2', 'ATP8', 'ATP6', 'COX3', 'ND3', 'ND4L', 'ND4', 'ND5', 'ND6', 'CYTB']) \n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in feature_dict.keys():\n",
    "    print(feature_dict[i].keys(), \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running clustal alingment for BDIP_ARG_vs_BPUB_RS\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">BDIP_ARG : 961\n",
      ">BPUB_RS : 961\n",
      "\n",
      "ND2.fa\n",
      ">BDIP_ARG : 1030\n",
      ">BPUB_RS : 1030\n",
      "\n",
      "COX1.fa\n",
      ">BDIP_ARG : 1602\n",
      ">BPUB_RS : 1602\n",
      "\n",
      "COX2.fa\n",
      ">BDIP_ARG : 685\n",
      ">BPUB_RS : 685\n",
      "\n",
      "ATP8.fa\n",
      ">BDIP_ARG : 165\n",
      ">BPUB_RS : 165\n",
      "\n",
      "ATP6.fa\n",
      ">BDIP_ARG : 680\n",
      ">BPUB_RS : 680\n",
      "\n",
      "COX3.fa\n",
      ">BDIP_ARG : 784\n",
      ">BPUB_RS : 784\n",
      "\n",
      "ND3.fa\n",
      ">BDIP_ARG : 343\n",
      ">BPUB_RS : 343\n",
      "\n",
      "ND4L.fa\n",
      ">BDIP_ARG : 290\n",
      ">BPUB_RS : 290\n",
      "\n",
      "ND4.fa\n",
      ">BDIP_ARG : 1338\n",
      ">BPUB_RS : 1338\n",
      "\n",
      "ND5.fa\n",
      ">BDIP_ARG : 1782\n",
      ">BPUB_RS : 1782\n",
      "\n",
      "ND6.fa\n",
      ">BDIP_ARG : 522\n",
      ">BPUB_RS : 522\n",
      "\n",
      "CYTB.fa\n",
      ">BDIP_ARG : 1114\n",
      ">BPUB_RS : 1114\n",
      "\n",
      "BDIP_ARG BPUB_RS\n",
      "11296 11296\n",
      "Features: 13, Alignment_size: 11296\n",
      "Running clustal alingment for BDIP_ARG_vs_Bothrops_jararaca\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">BDIP_ARG : 961\n",
      ">Bothrops_jararaca : 961\n",
      "\n",
      "ND2.fa\n",
      ">BDIP_ARG : 1030\n",
      ">Bothrops_jararaca : 1030\n",
      "\n",
      "COX1.fa\n",
      ">BDIP_ARG : 1602\n",
      ">Bothrops_jararaca : 1602\n",
      "\n",
      "COX2.fa\n",
      ">BDIP_ARG : 685\n",
      ">Bothrops_jararaca : 685\n",
      "\n",
      "ATP8.fa\n",
      ">BDIP_ARG : 165\n",
      ">Bothrops_jararaca : 165\n",
      "\n",
      "ATP6.fa\n",
      ">BDIP_ARG : 681\n",
      ">Bothrops_jararaca : 681\n",
      "\n",
      "COX3.fa\n",
      ">BDIP_ARG : 784\n",
      ">Bothrops_jararaca : 784\n",
      "\n",
      "ND3.fa\n",
      ">BDIP_ARG : 343\n",
      ">Bothrops_jararaca : 343\n",
      "\n",
      "ND4L.fa\n",
      ">BDIP_ARG : 291\n",
      ">Bothrops_jararaca : 291\n",
      "\n",
      "ND4.fa\n",
      ">BDIP_ARG : 1338\n",
      ">Bothrops_jararaca : 1338\n",
      "\n",
      "ND5.fa\n",
      ">BDIP_ARG : 1788\n",
      ">Bothrops_jararaca : 1788\n",
      "\n",
      "ND6.fa\n",
      ">BDIP_ARG : 522\n",
      ">Bothrops_jararaca : 522\n",
      "\n",
      "CYTB.fa\n",
      ">BDIP_ARG : 1114\n",
      ">Bothrops_jararaca : 1114\n",
      "\n",
      "BDIP_ARG Bothrops_jararaca\n",
      "11304 11304\n",
      "Features: 13, Alignment_size: 11304\n",
      "Running clustal alingment for BDIP_ARG_vs_Bothrops_pubescens\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">BDIP_ARG : 961\n",
      ">Bothrops_pubescens : 961\n",
      "\n",
      "ND2.fa\n",
      ">BDIP_ARG : 1030\n",
      ">Bothrops_pubescens : 1030\n",
      "\n",
      "COX1.fa\n",
      ">BDIP_ARG : 1602\n",
      ">Bothrops_pubescens : 1602\n",
      "\n",
      "COX2.fa\n",
      ">BDIP_ARG : 685\n",
      ">Bothrops_pubescens : 685\n",
      "\n",
      "ATP8.fa\n",
      ">BDIP_ARG : 186\n",
      ">Bothrops_pubescens : 186\n",
      "\n",
      "ATP6.fa\n",
      ">BDIP_ARG : 681\n",
      ">Bothrops_pubescens : 681\n",
      "\n",
      "COX3.fa\n",
      ">BDIP_ARG : 784\n",
      ">Bothrops_pubescens : 784\n",
      "\n",
      "ND3.fa\n",
      ">BDIP_ARG : 343\n",
      ">Bothrops_pubescens : 343\n",
      "\n",
      "ND4L.fa\n",
      ">BDIP_ARG : 291\n",
      ">Bothrops_pubescens : 291\n",
      "\n",
      "ND4.fa\n",
      ">BDIP_ARG : 1338\n",
      ">Bothrops_pubescens : 1338\n",
      "\n",
      "ND5.fa\n",
      ">BDIP_ARG : 1782\n",
      ">Bothrops_pubescens : 1782\n",
      "\n",
      "ND6.fa\n",
      ">BDIP_ARG : 522\n",
      ">Bothrops_pubescens : 522\n",
      "\n",
      "CYTB.fa\n",
      ">BDIP_ARG : 1114\n",
      ">Bothrops_pubescens : 1114\n",
      "\n",
      "BDIP_ARG Bothrops_pubescens\n",
      "11319 11319\n",
      "Features: 13, Alignment_size: 11319\n",
      "Running clustal alingment for BDIP_ARG_vs_Bothrops_diporus\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">BDIP_ARG : 961\n",
      ">Bothrops_diporus : 961\n",
      "\n",
      "ND2.fa\n",
      ">BDIP_ARG : 1030\n",
      ">Bothrops_diporus : 1030\n",
      "\n",
      "COX1.fa\n",
      ">BDIP_ARG : 1602\n",
      ">Bothrops_diporus : 1602\n",
      "\n",
      "COX2.fa\n",
      ">BDIP_ARG : 685\n",
      ">Bothrops_diporus : 685\n",
      "\n",
      "ATP8.fa\n",
      ">BDIP_ARG : 165\n",
      ">Bothrops_diporus : 165\n",
      "\n",
      "ATP6.fa\n",
      ">BDIP_ARG : 681\n",
      ">Bothrops_diporus : 681\n",
      "\n",
      "COX3.fa\n",
      ">BDIP_ARG : 784\n",
      ">Bothrops_diporus : 784\n",
      "\n",
      "ND3.fa\n",
      ">BDIP_ARG : 343\n",
      ">Bothrops_diporus : 343\n",
      "\n",
      "ND4L.fa\n",
      ">BDIP_ARG : 291\n",
      ">Bothrops_diporus : 291\n",
      "\n",
      "ND4.fa\n",
      ">BDIP_ARG : 1338\n",
      ">Bothrops_diporus : 1338\n",
      "\n",
      "ND5.fa\n",
      ">BDIP_ARG : 1782\n",
      ">Bothrops_diporus : 1782\n",
      "\n",
      "ND6.fa\n",
      ">BDIP_ARG : 522\n",
      ">Bothrops_diporus : 522\n",
      "\n",
      "CYTB.fa\n",
      ">BDIP_ARG : 1114\n",
      ">Bothrops_diporus : 1114\n",
      "\n",
      "BDIP_ARG Bothrops_diporus\n",
      "11298 11298\n",
      "Features: 13, Alignment_size: 11298\n",
      "Running clustal alingment for BPUB_RS_vs_Bothrops_jararaca\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">BPUB_RS : 961\n",
      ">Bothrops_jararaca : 961\n",
      "\n",
      "ND2.fa\n",
      ">BPUB_RS : 1030\n",
      ">Bothrops_jararaca : 1030\n",
      "\n",
      "COX1.fa\n",
      ">BPUB_RS : 1602\n",
      ">Bothrops_jararaca : 1602\n",
      "\n",
      "COX2.fa\n",
      ">BPUB_RS : 685\n",
      ">Bothrops_jararaca : 685\n",
      "\n",
      "ATP8.fa\n",
      ">BPUB_RS : 165\n",
      ">Bothrops_jararaca : 165\n",
      "\n",
      "ATP6.fa\n",
      ">BPUB_RS : 681\n",
      ">Bothrops_jararaca : 681\n",
      "\n",
      "COX3.fa\n",
      ">BPUB_RS : 784\n",
      ">Bothrops_jararaca : 784\n",
      "\n",
      "ND3.fa\n",
      ">BPUB_RS : 343\n",
      ">Bothrops_jararaca : 343\n",
      "\n",
      "ND4L.fa\n",
      ">BPUB_RS : 291\n",
      ">Bothrops_jararaca : 291\n",
      "\n",
      "ND4.fa\n",
      ">BPUB_RS : 1338\n",
      ">Bothrops_jararaca : 1338\n",
      "\n",
      "ND5.fa\n",
      ">BPUB_RS : 1788\n",
      ">Bothrops_jararaca : 1788\n",
      "\n",
      "ND6.fa\n",
      ">BPUB_RS : 522\n",
      ">Bothrops_jararaca : 522\n",
      "\n",
      "CYTB.fa\n",
      ">BPUB_RS : 1114\n",
      ">Bothrops_jararaca : 1114\n",
      "\n",
      "BPUB_RS Bothrops_jararaca\n",
      "11304 11304\n",
      "Features: 13, Alignment_size: 11304\n",
      "Running clustal alingment for BPUB_RS_vs_Bothrops_pubescens\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">BPUB_RS : 961\n",
      ">Bothrops_pubescens : 961\n",
      "\n",
      "ND2.fa\n",
      ">BPUB_RS : 1030\n",
      ">Bothrops_pubescens : 1030\n",
      "\n",
      "COX1.fa\n",
      ">BPUB_RS : 1602\n",
      ">Bothrops_pubescens : 1602\n",
      "\n",
      "COX2.fa\n",
      ">BPUB_RS : 685\n",
      ">Bothrops_pubescens : 685\n",
      "\n",
      "ATP8.fa\n",
      ">BPUB_RS : 186\n",
      ">Bothrops_pubescens : 186\n",
      "\n",
      "ATP6.fa\n",
      ">BPUB_RS : 681\n",
      ">Bothrops_pubescens : 681\n",
      "\n",
      "COX3.fa\n",
      ">BPUB_RS : 784\n",
      ">Bothrops_pubescens : 784\n",
      "\n",
      "ND3.fa\n",
      ">BPUB_RS : 343\n",
      ">Bothrops_pubescens : 343\n",
      "\n",
      "ND4L.fa\n",
      ">BPUB_RS : 291\n",
      ">Bothrops_pubescens : 291\n",
      "\n",
      "ND4.fa\n",
      ">BPUB_RS : 1338\n",
      ">Bothrops_pubescens : 1338\n",
      "\n",
      "ND5.fa\n",
      ">BPUB_RS : 1782\n",
      ">Bothrops_pubescens : 1782\n",
      "\n",
      "ND6.fa\n",
      ">BPUB_RS : 522\n",
      ">Bothrops_pubescens : 522\n",
      "\n",
      "CYTB.fa\n",
      ">BPUB_RS : 1114\n",
      ">Bothrops_pubescens : 1114\n",
      "\n",
      "BPUB_RS Bothrops_pubescens\n",
      "11319 11319\n",
      "Features: 13, Alignment_size: 11319\n",
      "Running clustal alingment for BPUB_RS_vs_Bothrops_diporus\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">BPUB_RS : 961\n",
      ">Bothrops_diporus : 961\n",
      "\n",
      "ND2.fa\n",
      ">BPUB_RS : 1030\n",
      ">Bothrops_diporus : 1030\n",
      "\n",
      "COX1.fa\n",
      ">BPUB_RS : 1602\n",
      ">Bothrops_diporus : 1602\n",
      "\n",
      "COX2.fa\n",
      ">BPUB_RS : 685\n",
      ">Bothrops_diporus : 685\n",
      "\n",
      "ATP8.fa\n",
      ">BPUB_RS : 165\n",
      ">Bothrops_diporus : 165\n",
      "\n",
      "ATP6.fa\n",
      ">BPUB_RS : 681\n",
      ">Bothrops_diporus : 681\n",
      "\n",
      "COX3.fa\n",
      ">BPUB_RS : 784\n",
      ">Bothrops_diporus : 784\n",
      "\n",
      "ND3.fa\n",
      ">BPUB_RS : 343\n",
      ">Bothrops_diporus : 343\n",
      "\n",
      "ND4L.fa\n",
      ">BPUB_RS : 291\n",
      ">Bothrops_diporus : 291\n",
      "\n",
      "ND4.fa\n",
      ">BPUB_RS : 1338\n",
      ">Bothrops_diporus : 1338\n",
      "\n",
      "ND5.fa\n",
      ">BPUB_RS : 1782\n",
      ">Bothrops_diporus : 1782\n",
      "\n",
      "ND6.fa\n",
      ">BPUB_RS : 522\n",
      ">Bothrops_diporus : 522\n",
      "\n",
      "CYTB.fa\n",
      ">BPUB_RS : 1114\n",
      ">Bothrops_diporus : 1114\n",
      "\n",
      "BPUB_RS Bothrops_diporus\n",
      "11298 11298\n",
      "Features: 13, Alignment_size: 11298\n",
      "Running clustal alingment for Bothrops_jararaca_vs_Bothrops_pubescens\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">Bothrops_jararaca : 961\n",
      ">Bothrops_pubescens : 961\n",
      "\n",
      "ND2.fa\n",
      ">Bothrops_jararaca : 1030\n",
      ">Bothrops_pubescens : 1030\n",
      "\n",
      "COX1.fa\n",
      ">Bothrops_jararaca : 1602\n",
      ">Bothrops_pubescens : 1602\n",
      "\n",
      "COX2.fa\n",
      ">Bothrops_jararaca : 685\n",
      ">Bothrops_pubescens : 685\n",
      "\n",
      "ATP8.fa\n",
      ">Bothrops_jararaca : 186\n",
      ">Bothrops_pubescens : 186\n",
      "\n",
      "ATP6.fa\n",
      ">Bothrops_jararaca : 681\n",
      ">Bothrops_pubescens : 681\n",
      "\n",
      "COX3.fa\n",
      ">Bothrops_jararaca : 784\n",
      ">Bothrops_pubescens : 784\n",
      "\n",
      "ND3.fa\n",
      ">Bothrops_jararaca : 343\n",
      ">Bothrops_pubescens : 343\n",
      "\n",
      "ND4L.fa\n",
      ">Bothrops_jararaca : 291\n",
      ">Bothrops_pubescens : 291\n",
      "\n",
      "ND4.fa\n",
      ">Bothrops_jararaca : 1338\n",
      ">Bothrops_pubescens : 1338\n",
      "\n",
      "ND5.fa\n",
      ">Bothrops_jararaca : 1788\n",
      ">Bothrops_pubescens : 1788\n",
      "\n",
      "ND6.fa\n",
      ">Bothrops_jararaca : 522\n",
      ">Bothrops_pubescens : 522\n",
      "\n",
      "CYTB.fa\n",
      ">Bothrops_jararaca : 1114\n",
      ">Bothrops_pubescens : 1114\n",
      "\n",
      "Bothrops_jararaca Bothrops_pubescens\n",
      "11325 11325\n",
      "Features: 13, Alignment_size: 11325\n",
      "Running clustal alingment for Bothrops_jararaca_vs_Bothrops_diporus\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">Bothrops_jararaca : 961\n",
      ">Bothrops_diporus : 961\n",
      "\n",
      "ND2.fa\n",
      ">Bothrops_jararaca : 1030\n",
      ">Bothrops_diporus : 1030\n",
      "\n",
      "COX1.fa\n",
      ">Bothrops_jararaca : 1602\n",
      ">Bothrops_diporus : 1602\n",
      "\n",
      "COX2.fa\n",
      ">Bothrops_jararaca : 685\n",
      ">Bothrops_diporus : 685\n",
      "\n",
      "ATP8.fa\n",
      ">Bothrops_jararaca : 165\n",
      ">Bothrops_diporus : 165\n",
      "\n",
      "ATP6.fa\n",
      ">Bothrops_jararaca : 681\n",
      ">Bothrops_diporus : 681\n",
      "\n",
      "COX3.fa\n",
      ">Bothrops_jararaca : 784\n",
      ">Bothrops_diporus : 784\n",
      "\n",
      "ND3.fa\n",
      ">Bothrops_jararaca : 343\n",
      ">Bothrops_diporus : 343\n",
      "\n",
      "ND4L.fa\n",
      ">Bothrops_jararaca : 291\n",
      ">Bothrops_diporus : 291\n",
      "\n",
      "ND4.fa\n",
      ">Bothrops_jararaca : 1338\n",
      ">Bothrops_diporus : 1338\n",
      "\n",
      "ND5.fa\n",
      ">Bothrops_jararaca : 1788\n",
      ">Bothrops_diporus : 1788\n",
      "\n",
      "ND6.fa\n",
      ">Bothrops_jararaca : 522\n",
      ">Bothrops_diporus : 522\n",
      "\n",
      "CYTB.fa\n",
      ">Bothrops_jararaca : 1114\n",
      ">Bothrops_diporus : 1114\n",
      "\n",
      "Bothrops_jararaca Bothrops_diporus\n",
      "11304 11304\n",
      "Features: 13, Alignment_size: 11304\n",
      "Running clustal alingment for Bothrops_pubescens_vs_Bothrops_diporus\n",
      "Replacing ambiguous sites with gap characters...\n",
      "ND1.fa\n",
      ">Bothrops_pubescens : 961\n",
      ">Bothrops_diporus : 961\n",
      "\n",
      "ND2.fa\n",
      ">Bothrops_pubescens : 1030\n",
      ">Bothrops_diporus : 1030\n",
      "\n",
      "COX1.fa\n",
      ">Bothrops_pubescens : 1602\n",
      ">Bothrops_diporus : 1602\n",
      "\n",
      "COX2.fa\n",
      ">Bothrops_pubescens : 685\n",
      ">Bothrops_diporus : 685\n",
      "\n",
      "ATP8.fa\n",
      ">Bothrops_pubescens : 186\n",
      ">Bothrops_diporus : 186\n",
      "\n",
      "ATP6.fa\n",
      ">Bothrops_pubescens : 681\n",
      ">Bothrops_diporus : 681\n",
      "\n",
      "COX3.fa\n",
      ">Bothrops_pubescens : 784\n",
      ">Bothrops_diporus : 784\n",
      "\n",
      "ND3.fa\n",
      ">Bothrops_pubescens : 343\n",
      ">Bothrops_diporus : 343\n",
      "\n",
      "ND4L.fa\n",
      ">Bothrops_pubescens : 291\n",
      ">Bothrops_diporus : 291\n",
      "\n",
      "ND4.fa\n",
      ">Bothrops_pubescens : 1338\n",
      ">Bothrops_diporus : 1338\n",
      "\n",
      "ND5.fa\n",
      ">Bothrops_pubescens : 1782\n",
      ">Bothrops_diporus : 1782\n",
      "\n",
      "ND6.fa\n",
      ">Bothrops_pubescens : 522\n",
      ">Bothrops_diporus : 522\n",
      "\n",
      "CYTB.fa\n",
      ">Bothrops_pubescens : 1114\n",
      ">Bothrops_diporus : 1114\n",
      "\n",
      "Bothrops_pubescens Bothrops_diporus\n",
      "11319 11319\n",
      "Features: 13, Alignment_size: 11319\n"
     ]
    }
   ],
   "source": [
    "def create_folder(sp1, sp2):\n",
    "    folder_name = \"{}_vs_{}\".format(sp1, sp2)\n",
    "    os.mkdir(folder_name)\n",
    "    os.chdir(folder_name)\n",
    "    print(\"Running clustal alingment for {}\".format(folder_name))\n",
    "\n",
    "def run_clustalo(sp1, sp2, feature_dict):\n",
    "    sp1_feat = feature_dict[sp1]\n",
    "    sp2_feat = feature_dict[sp2]\n",
    "    clustal_list = list()\n",
    "    for i in sp1_feat.keys():\n",
    "        clustal_output = \"{}.fa\".format(i)\n",
    "        clustal_error = \"{}.err\".format(i)\n",
    "        with open(\"tempfile\", \"w+\") as tempfile, open(clustal_output, \"w+\") as clustal_out, open(clustal_error, \"w+\") as clustal_err:\n",
    "            tempfile.write(\">{}\\n{}\\n>{}\\n{}\".format(sp1, sp1_feat[i], sp2, sp2_feat[i]))\n",
    "            tempfile.close()\n",
    "            #time.sleep(2)\n",
    "            alignment = subprocess.Popen([\"clustalo\", \"-i\", \"tempfile\"], stdout=clustal_out, stderr=clustal_err)\n",
    "            alignment.wait()\n",
    "            os.remove(\"tempfile\")\n",
    "        if os.path.getsize(clustal_output) != 0 and os.path.getsize(clustal_error) == 0:\n",
    "            os.remove(clustal_error)\n",
    "            clustal_list.append(clustal_output)\n",
    "        else:\n",
    "            print(\"There was a problem with the {}_vs{} alignment. Check the file {}\".format(sp1, sp2, clustal_error))\n",
    "    return(clustal_list)\n",
    "\n",
    "def ambiguous_sites_as_gaps(clustal_list):\n",
    "    print(\"Replacing ambiguous sites with gap characters...\")\n",
    "    data = str()\n",
    "    for aln in clustal_list:\n",
    "        data = str()\n",
    "        data2 = str()\n",
    "        print(aln)\n",
    "        alignment = AlignIO.read(aln, \"fasta\")\n",
    "        for index in [0, 1]: ##Index of each fasta record in the alignment - in this case, 0 and 1 (pairwise)\n",
    "            valid_nucs = [\"A\", \"T\", \"G\", \"C\"]\n",
    "            unambiguous_seq = str()\n",
    "            for nuc in alignment[index].seq:\n",
    "                if nuc in valid_nucs:\n",
    "                    unambiguous_seq += nuc\n",
    "                else:\n",
    "                    unambiguous_seq += \"-\"\n",
    "            data += \">{} : {}\\n\".format(alignment[index].id, len(unambiguous_seq))\n",
    "            data2 += \">{}\\n{}\\n\".format(alignment[index].id, unambiguous_seq)\n",
    "        print(data)\n",
    "        with open(aln, \"w\") as unambiguous_aln:\n",
    "            unambiguous_aln.write(data2)\n",
    "\n",
    "def concat_alignment(clustal_list):\n",
    "    ##Besides the merging, this function needs to generate a clustal list with all the absolute paths to\n",
    "    ##the final alignment files, to be used as input to the \"alignment_parser function\"\n",
    "    header1 = set()\n",
    "    header2 = set()\n",
    "    seq1 = \"\"\n",
    "    seq2 = \"\"\n",
    "    size = 0\n",
    "    number_of_features = 0\n",
    "    concat_filename = \"\"\n",
    "    for num, i in enumerate(clustal_list, 1):\n",
    "        aln = AlignIO.read(i, \"fasta\")\n",
    "        header1.add(aln[0].id)\n",
    "        header2.add(aln[1].id)\n",
    "        seq1 += aln[0].seq\n",
    "        seq2 += aln[1].seq\n",
    "        size += aln.get_alignment_length()\n",
    "        number_of_features = num\n",
    "    if len(header1) == len(header2) == 1 and size == len(seq1) == len(seq2):\n",
    "        header1 = list(header1)[0]\n",
    "        header2 = list(header2)[0]\n",
    "        concat_filename = \"{}_vs_{}_concat.fa\".format(header1, header2)\n",
    "        print(header1, header2)\n",
    "        print(len(seq1), len(seq2))\n",
    "        print(\"Features: {}, Alignment_size: {}\".format(number_of_features, size))\n",
    "        with open(concat_filename, \"w\") as concat:\n",
    "            concat.write(\">{}\\n{}\\n>{}\\n{}\\n\".format(header1, seq1, header2, seq2))\n",
    "    else:\n",
    "        print(\"There was a problem in the merging of alignments in {}\".format(concat_filename))\n",
    "        return(None)\n",
    "    return(os.path.abspath(concat_filename))\n",
    "    \n",
    "def pairwise_permutations(feature_dict):\n",
    "    '''Receives a dictionary of multiple sequence entries (e.g. {header1 : seq1, header2 : seq2, ... headerN : seqN}) and returns a list of all possible pairwise permutations between the dictionary's keys.'''\n",
    "    permutations = list()\n",
    "    global concat_list\n",
    "    concat_list = list()\n",
    "    for sp1, sp2 in itertools.combinations(feature_dict.keys(), 2):\n",
    "        create_folder(sp1, sp2)\n",
    "        clustal_list = run_clustalo(sp1, sp2, feature_dict)\n",
    "        ambiguous_sites_as_gaps(clustal_list)\n",
    "        concat_list.append(concat_alignment(clustal_list))\n",
    "        os.chdir(\"..\")        \n",
    "        \n",
    "pairwise_permutations(feature_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/BDIP_ARG_vs_BPUB_RS/BDIP_ARG_vs_BPUB_RS_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/BDIP_ARG_vs_Bothrops_jararaca/BDIP_ARG_vs_Bothrops_jararaca_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/BDIP_ARG_vs_Bothrops_pubescens/BDIP_ARG_vs_Bothrops_pubescens_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/BDIP_ARG_vs_Bothrops_diporus/BDIP_ARG_vs_Bothrops_diporus_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/BPUB_RS_vs_Bothrops_jararaca/BPUB_RS_vs_Bothrops_jararaca_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/BPUB_RS_vs_Bothrops_pubescens/BPUB_RS_vs_Bothrops_pubescens_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/BPUB_RS_vs_Bothrops_diporus/BPUB_RS_vs_Bothrops_diporus_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/Bothrops_jararaca_vs_Bothrops_pubescens/Bothrops_jararaca_vs_Bothrops_pubescens_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/Bothrops_jararaca_vs_Bothrops_diporus/Bothrops_jararaca_vs_Bothrops_diporus_concat.fa',\n",
       " '/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_CDS/mitos/Bothrops_pubescens_vs_Bothrops_diporus/Bothrops_pubescens_vs_Bothrops_diporus_concat.fa']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clustal_parser(concat_list):\n",
    "    fields = [\"Query MT\", \"Subject MT\", \"Total Alignment Sites\", \"Gapped Sites\", \"Valid Sites\", \"Identical Sites\", \"Variant Sites\", \"Identity Percentage\"]\n",
    "    excel_output_dict = {k : list() for k in fields}\n",
    "    for file in concat_list: ##First, we need to get each information from the alignment and assign it to a variable\n",
    "        aln = AlignIO.read(file, \"fasta\")\n",
    "        query_seq = aln[0].id\n",
    "        subject_seq = aln[1].id\n",
    "        total_aln_sites = aln.get_alignment_length()\n",
    "        gap_sites = list()\n",
    "        gap_sites_count = 0\n",
    "        identical_sites = list()\n",
    "        identical_sites_count = 0\n",
    "        variant_sites = list()\n",
    "        variant_sites_count = 0\n",
    "        collumns = dict()\n",
    "        for i in range(1, total_aln_sites + 1):\n",
    "            collumns.update({i : aln[:, i-1]})\n",
    "        for k, v in collumns.items(): ##Get number of gapped sites\n",
    "            if \"-\" in v:\n",
    "                gap_sites_count += 1\n",
    "                gap_sites.append([k, v])\n",
    "                #print(k, v)\n",
    "                continue\n",
    "            nucs_in_collumn = set()\n",
    "            for i in [0,1]: ##Get number of identical and variable sites for pairwise alignment\n",
    "                nucs_in_collumn.update([v[i]])\n",
    "            if len(nucs_in_collumn) == 1:\n",
    "                identical_sites_count += 1\n",
    "                identical_sites.append([k, v])\n",
    "            if len(nucs_in_collumn) > 1:\n",
    "                variant_sites_count += 1\n",
    "                variant_sites.append([k, v])\n",
    "                #print(k, v)        \n",
    "        valid_sites = total_aln_sites - gap_sites_count\n",
    "        identity_percentage = round((identical_sites_count / valid_sites) * 100, 2)\n",
    "        \n",
    "        \n",
    "        ##After getting the variables, the script generates a file for each pairwise alignment, with a summary of the alignment's data\n",
    "        \n",
    "        parser_output = \"{}.clustalparser\".format(file[:-3])        \n",
    "        with open(parser_output, \"w+\") as cluparser:\n",
    "            cluparser.write(\"Query sequence: {} - A {} \\t T {} \\t G {} \\t C {} \\t gaps {}\\n\".format(query_seq, aln[0].seq.count(\"A\"), aln[0].seq.count(\"T\"), aln[0].seq.count(\"G\"), aln[0].seq.count(\"C\"), aln[0].seq.count(\"-\")))\n",
    "            cluparser.write(\"Subject sequence: {} - A {} \\t T {} \\t G {} \\t C {} \\t gaps {}\\n\".format(subject_seq, aln[1].seq.count(\"A\"), aln[1].seq.count(\"T\"), aln[1].seq.count(\"G\"), aln[1].seq.count(\"C\"), aln[1].seq.count(\"-\")))\n",
    "            cluparser.write(\"Total alignment size: {}\\n\".format(total_aln_sites))\n",
    "            cluparser.write(\"Gapped columns: {}\\n\".format(gap_sites_count))\n",
    "            cluparser.write(\"Valid columns (Total - Gapped): {}\\n\".format(valid_sites))\n",
    "            cluparser.write(\"Identical columns: {}\\n\".format(identical_sites_count))\n",
    "            cluparser.write(\"Variant columns: {}\\n\".format(variant_sites_count))\n",
    "            cluparser.write(\"Identity percentage - ((Identical collumns / Valid collumns) * 100): {} %\\n\".format(identity_percentage))\n",
    "            cluparser.write(\"VARIANT SITES:\\n\")\n",
    "            cluparser.write(\"Number\\tAlignment_site\\t{}_nt\\t{}_nt\\n\".format(query_seq, subject_seq))\n",
    "            [cluparser.write(\"{}\\t{}\\t{}\\t{}\\n\".format(i, v[0], v[1][0], v[1][1])) for i, v in enumerate(variant_sites, 1)]\n",
    "            cluparser.write(\"GAP SITES:\\n\")\n",
    "            cluparser.write(\"Number\\tAlignment_site\\t{}_nt\\t{}_nt\\n\".format(query_seq, subject_seq))\n",
    "            [cluparser.write(\"{}\\t{}\\t{}\\t{}\\n\".format(i, v[0], v[1][0], v[1][1])) for i, v in enumerate(gap_sites, 1)]\n",
    "            cluparser.write(\"IDENTICAL SITES:\\n\")\n",
    "            cluparser.write(\"Number\\tAlignment_site\\t{}_nt\\t{}_nt\\n\".format(query_seq, subject_seq))\n",
    "            [cluparser.write(\"{}\\t{}\\t{}\\t{}\\n\".format(i, v[0], v[1][0], v[1][1])) for i, v in enumerate(identical_sites, 1)]\n",
    "\n",
    "        ##Lastly, the script will save all data to a dictionary and return it. This dict will be used to create a pandas dataframe and write a excel spreadsheet\n",
    "            \n",
    "        excel_output_dict.get(\"Query MT\").append(query_seq)\n",
    "        excel_output_dict.get(\"Subject MT\").append(subject_seq)\n",
    "        excel_output_dict.get(\"Total Alignment Sites\").append(total_aln_sites)\n",
    "        excel_output_dict.get(\"Gapped Sites\").append(gap_sites_count)\n",
    "        excel_output_dict.get(\"Valid Sites\").append(valid_sites)\n",
    "        excel_output_dict.get(\"Identical Sites\").append(identical_sites_count)\n",
    "        excel_output_dict.get(\"Variant Sites\").append(variant_sites_count)\n",
    "        excel_output_dict.get(\"Identity Percentage\").append(identity_percentage)\n",
    "            \n",
    "        #if (gap_sites_count + identical_sites_count + variant_sites_count) == total_aln_sites:\n",
    "         #   print(\"The sum is correct\")\n",
    "        #else:\n",
    "        #    print(\"There was a problem with the parsing\")\n",
    "        #print(aln[1].seq.count(\"-\"))\n",
    "        \n",
    "    return(excel_output_dict)\n",
    "\n",
    "excel = clustal_parser(concat_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_excel_aln_summary(excel_output_dict):\n",
    "    aln_summary = pandas.DataFrame(excel_output_dict)\n",
    "    aln_summary.to_excel(\"Pairwise_alignment_summary.xlsx\", index=False, sheet_name=\"Aln_summary\")   \n",
    "\n",
    "generate_excel_aln_summary(excel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/ultratb.py\", line 1132, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/ultratb.py\", line 313, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/ultratb.py\", line 358, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/inspect.py\", line 1058, in getinnerframes\n",
      "    framelist.append((tb.tb_frame,) + getframeinfo(tb, context))\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/inspect.py\", line 1018, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/inspect.py\", line 453, in getsourcefile\n",
      "    if hasattr(getmodule(object, filename), '__loader__'):\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/inspect.py\", line 482, in getmodule\n",
      "    file = getabsfile(object, _filename)\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/inspect.py\", line 466, in getabsfile\n",
      "    return os.path.normcase(os.path.abspath(_filename))\n",
      "  File \"/home/gabriel/miniconda2/lib/python2.7/posixpath.py\", line 371, in abspath\n",
      "    cwd = os.getcwd()\n",
      "OSError: [Errno 2] No such file or directory\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "string index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc\u001b[0m in \u001b[0;36mrun_code\u001b[0;34m(self, code_obj, result)\u001b[0m\n\u001b[1;32m   2893\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2894\u001b[0m                 \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror_in_exec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexc_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2895\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshowtraceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2896\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2897\u001b[0m             \u001b[0moutflag\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.pyc\u001b[0m in \u001b[0;36mshowtraceback\u001b[0;34m(self, exc_tuple, filename, tb_offset, exception_only)\u001b[0m\n\u001b[1;32m   1824\u001b[0m                     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1825\u001b[0m                         stb = self.InteractiveTB.structured_traceback(etype,\n\u001b[0;32m-> 1826\u001b[0;31m                                             value, tb, tb_offset=tb_offset)\n\u001b[0m\u001b[1;32m   1827\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1828\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_showtraceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/ultratb.pyc\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1409\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1410\u001b[0m         return FormattedTB.structured_traceback(\n\u001b[0;32m-> 1411\u001b[0;31m             self, etype, value, tb, tb_offset, number_of_lines_of_context)\n\u001b[0m\u001b[1;32m   1412\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1413\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/ultratb.pyc\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1317\u001b[0m             \u001b[0;31m# Verbose modes need a full traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1318\u001b[0m             return VerboseTB.structured_traceback(\n\u001b[0;32m-> 1319\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb_offset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumber_of_lines_of_context\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1320\u001b[0m             )\n\u001b[1;32m   1321\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/home/gabriel/miniconda2/lib/python2.7/site-packages/IPython/core/ultratb.pyc\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1202\u001b[0m                 \u001b[0mstructured_traceback_parts\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mformatted_exception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1203\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1204\u001b[0;31m             \u001b[0mstructured_traceback_parts\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mformatted_exception\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1206\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mstructured_traceback_parts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: string index out of range"
     ]
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/gabriel/Dropbox/repos/python_bioinfo/pairwise_align_features'"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"Bothrops_pubescens_vs_Bothrops_diporus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['tRNA-Arg.fa',\n",
       " 'ATP8.fa',\n",
       " 'tRNA-Ile.fa',\n",
       " 'tRNA-Ser2.fa',\n",
       " 'tRNA-Met.fa',\n",
       " 'CYTB.fa',\n",
       " 'tRNA-Phe.fa',\n",
       " 'tRNA-Asn.fa',\n",
       " 'ATP6.fa',\n",
       " 'tRNA-Pro.fa',\n",
       " 'tRNA-Cys.fa',\n",
       " 'tRNA-Asp.fa',\n",
       " 'tRNA-Thr.fa',\n",
       " 'tRNA-Gly.fa',\n",
       " 'tRNA-Ala.fa',\n",
       " 'tRNA-Tyr.fa',\n",
       " 'tRNA-Ser.fa',\n",
       " 'tRNA-Gln.fa',\n",
       " 'tRNA-Leu.fa',\n",
       " 'COX2.fa',\n",
       " 'tRNA-Trp.fa',\n",
       " 'tRNA-Leu2.fa',\n",
       " 'COX1.fa',\n",
       " 'tRNA-His.fa',\n",
       " 'tRNA-Val.fa',\n",
       " 'tRNA-Glu.fa',\n",
       " 'ND4L.fa',\n",
       " '12S.fa',\n",
       " 'tRNA-Lys.fa',\n",
       " 'COX3.fa',\n",
       " 'ND1.fa',\n",
       " 'ND6.fa',\n",
       " 'ND3.fa',\n",
       " 'ND2.fa',\n",
       " 'ND5.fa',\n",
       " 'ND4.fa',\n",
       " '16S.fa']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clustal_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Bothrops_pubescens', 'Bothrops_diporus')\n",
      "(15146, 15146)\n",
      "Features: 37, Alignment_size: 15146\n"
     ]
    }
   ],
   "source": [
    "def concat_alignments(clustal_list):\n",
    "    ##Besides the merging, this function needs to generate a clustal list with all the absolute paths to\n",
    "    ##the final alignment files, to be used as input to the \"alignment_parser function\"\n",
    "    header1 = set()\n",
    "    header2 = set()\n",
    "    seq1 = \"\"\n",
    "    seq2 = \"\"\n",
    "    size = 0\n",
    "    number_of_features = 0\n",
    "    concat_filename = \"\"\n",
    "    for num, i in enumerate(clustal_list, 1):\n",
    "        aln = AlignIO.read(i, \"fasta\")\n",
    "        header1.add(aln[0].id)\n",
    "        header2.add(aln[1].id)\n",
    "        seq1 += aln[0].seq\n",
    "        seq2 += aln[1].seq\n",
    "        size += aln.get_alignment_length()\n",
    "        number_of_features = num\n",
    "    if len(header1) == len(header2) == 1 and size == len(seq1) == len(seq2):\n",
    "        header1 = list(header1)[0]\n",
    "        header2 = list(header2)[0]\n",
    "        concat_filename = \"{}_vs_{}_concat.fa\".format(header1, header2)\n",
    "        print(header1, header2)\n",
    "        print(len(seq1), len(seq2))\n",
    "        print(\"Features: {}, Alignment_size: {}\".format(number_of_features, size))\n",
    "        with open(concat_filename, \"w\") as concat:\n",
    "            concat.write(\">{}\\n{}\\n>{}\\n{}\\n\".format(header1, seq1, header2, seq2))\n",
    "    else:\n",
    "        print(\"There was a problem in the merging of alignments in {}\".format(concat_filename))\n",
    "        return(None)\n",
    "    return(concat_filename)\n",
    "\n",
    "concat = concat_alignments(clustal_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Bothrops_pubescens_vs_Bothrops_diporus_concat.fa']\n"
     ]
    }
   ],
   "source": [
    "concat_list = list()\n",
    "concat_list.append(concat)\n",
    "print(concat_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clustal_parser(concat_list):\n",
    "    fields = [\"Query MT\", \"Subject MT\", \"Total Alignment Sites\", \"Gapped Sites\", \"Valid Sites\", \"Identical Sites\", \"Variant Sites\", \"Identity Percentage\"]\n",
    "    excel_output_dict = {k : list() for k in fields}\n",
    "    for file in concat_list: ##First, we need to get each information from the alignment and assign it to a variable\n",
    "        aln = AlignIO.read(file, \"fasta\")\n",
    "        query_seq = aln[0].id\n",
    "        subject_seq = aln[1].id\n",
    "        total_aln_sites = aln.get_alignment_length()\n",
    "        gap_sites = list()\n",
    "        gap_sites_count = 0\n",
    "        identical_sites = list()\n",
    "        identical_sites_count = 0\n",
    "        variant_sites = list()\n",
    "        variant_sites_count = 0\n",
    "        collumns = dict()\n",
    "        for i in range(1, total_aln_sites + 1):\n",
    "            collumns.update({i : aln[:, i-1]})\n",
    "        for k, v in collumns.items(): ##Get number of gapped sites\n",
    "            if \"-\" in v:\n",
    "                gap_sites_count += 1\n",
    "                gap_sites.append([k, v])\n",
    "                #print(k, v)\n",
    "                continue\n",
    "            nucs_in_collumn = set()\n",
    "            for i in [0,1]: ##Get number of identical and variable sites for pairwise alignment\n",
    "                nucs_in_collumn.update([v[i]])\n",
    "            if len(nucs_in_collumn) == 1:\n",
    "                identical_sites_count += 1\n",
    "                identical_sites.append([k, v])\n",
    "            if len(nucs_in_collumn) > 1:\n",
    "                variant_sites_count += 1\n",
    "                variant_sites.append([k, v])\n",
    "                #print(k, v)        \n",
    "        valid_sites = total_aln_sites - gap_sites_count\n",
    "        identity_percentage = round((identical_sites_count / valid_sites) * 100, 2)\n",
    "        \n",
    "        \n",
    "        ##After getting the variables, the script generates a file for each pairwise alignment, with a summary of the alignment's data\n",
    "        \n",
    "        parser_output = \"{}.clustalparser\".format(file[:-3])        \n",
    "        with open(parser_output, \"w+\") as cluparser:\n",
    "            cluparser.write(\"Query sequence: {} - A {} \\t T {} \\t G {} \\t C {} \\t gaps {}\\n\".format(query_seq, aln[0].seq.count(\"A\"), aln[0].seq.count(\"T\"), aln[0].seq.count(\"G\"), aln[0].seq.count(\"C\"), aln[0].seq.count(\"-\")))\n",
    "            cluparser.write(\"Subject sequence: {} - A {} \\t T {} \\t G {} \\t C {} \\t gaps {}\\n\".format(subject_seq, aln[1].seq.count(\"A\"), aln[1].seq.count(\"T\"), aln[1].seq.count(\"G\"), aln[1].seq.count(\"C\"), aln[1].seq.count(\"-\")))\n",
    "            cluparser.write(\"Total alignment size: {}\\n\".format(total_aln_sites))\n",
    "            cluparser.write(\"Gapped columns: {}\\n\".format(gap_sites_count))\n",
    "            cluparser.write(\"Valid columns (Total - Gapped): {}\\n\".format(valid_sites))\n",
    "            cluparser.write(\"Identical columns: {}\\n\".format(identical_sites_count))\n",
    "            cluparser.write(\"Variant columns: {}\\n\".format(variant_sites_count))\n",
    "            cluparser.write(\"Identity percentage - ((Identical collumns / Valid collumns) * 100): {} %\\n\".format(identity_percentage))\n",
    "            cluparser.write(\"VARIANT SITES:\\n\")\n",
    "            cluparser.write(\"Number\\tAlignment_site\\t{}_nt\\t{}_nt\\n\".format(query_seq, subject_seq))\n",
    "            [cluparser.write(\"{}\\t{}\\t{}\\t{}\\n\".format(i, v[0], v[1][0], v[1][1])) for i, v in enumerate(variant_sites, 1)]\n",
    "            cluparser.write(\"GAP SITES:\\n\")\n",
    "            cluparser.write(\"Number\\tAlignment_site\\t{}_nt\\t{}_nt\\n\".format(query_seq, subject_seq))\n",
    "            [cluparser.write(\"{}\\t{}\\t{}\\t{}\\n\".format(i, v[0], v[1][0], v[1][1])) for i, v in enumerate(gap_sites, 1)]\n",
    "            cluparser.write(\"IDENTICAL SITES:\\n\")\n",
    "            cluparser.write(\"Number\\tAlignment_site\\t{}_nt\\t{}_nt\\n\".format(query_seq, subject_seq))\n",
    "            [cluparser.write(\"{}\\t{}\\t{}\\t{}\\n\".format(i, v[0], v[1][0], v[1][1])) for i, v in enumerate(identical_sites, 1)]\n",
    "\n",
    "        ##Lastly, the script will save all data to a dictionary and return it. This dict will be used to create a pandas dataframe and write a excel spreadsheet\n",
    "            \n",
    "        excel_output_dict.get(\"Query MT\").append(query_seq)\n",
    "        excel_output_dict.get(\"Subject MT\").append(subject_seq)\n",
    "        excel_output_dict.get(\"Total Alignment Sites\").append(total_aln_sites)\n",
    "        excel_output_dict.get(\"Gapped Sites\").append(gap_sites_count)\n",
    "        excel_output_dict.get(\"Valid Sites\").append(valid_sites)\n",
    "        excel_output_dict.get(\"Identical Sites\").append(identical_sites_count)\n",
    "        excel_output_dict.get(\"Variant Sites\").append(variant_sites_count)\n",
    "        excel_output_dict.get(\"Identity Percentage\").append(identity_percentage)\n",
    "            \n",
    "        #if (gap_sites_count + identical_sites_count + variant_sites_count) == total_aln_sites:\n",
    "         #   print(\"The sum is correct\")\n",
    "        #else:\n",
    "        #    print(\"There was a problem with the parsing\")\n",
    "        #print(aln[1].seq.count(\"-\"))\n",
    "        \n",
    "    return(excel_output_dict)\n",
    "\n",
    "excel = clustal_parser(concat_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_excel_aln_summary(excel_output_dict):\n",
    "    aln_summary = pandas.DataFrame(excel_output_dict)\n",
    "    aln_summary.to_excel(\"Pairwise_alignment_summary.xlsx\", index=False, sheet_name=\"Aln_summary\")   \n",
    "\n",
    "generate_excel_aln_summary(excel)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What to do from here:\n",
    "\n",
    "- [x] Create the pairwise comparison list using the function from the other script\n",
    "- [x] For each pairwise comparison, create a folder\n",
    "- [x] Inside each folder, there will be an fasta for each gene pair\n",
    "- [x] Each fasta will be used for an alignment\n",
    "- [ ] Lastly, the sequece of all alignments will be merged into a single supermatrix (could be in any order, maybe the synteny order?)\n",
    "- [ ] This supermatrix will be used as input for the clustal_parser function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
